Allow Future::then to return pre-extracted DataPtrs

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* **#58424 Allow Future::then to return pre-extracted DataPtrs**
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

lw committed:
Allow Future::then to return pre-extracted DataPtrs

In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -508,20 +508,34 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
+    addCallback([childFut,
+                 cb = std::move(callback)](Future& parentFut) mutable {
+      try {
+        guts::if_constexpr<std::is_convertible<
```
curious, what's the difference between `guts::if_constexpr` and `if constexpr`? Is it because we are not using C++17 yet?

lw replied:
Correct. In fact, `guts::if_constexpr` will use `if constexpr` internally if available, but will fall back onto a C++-14-compatible polyfill if not.

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -508,20 +508,34 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
+        guts::if_constexpr<std::is_convertible<
+            typename std::result_of<T && (Future&)>::type,
+            IValueWithDataPtrs>::value>(
+            [&](auto _) {
```
any reason for using "_" instead of giving it a name?

lw replied:
I did it because this was the example in the docstring of `guts::if_constexpr`. I'll rename it to `identity`, hopefully that's clearer.

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -508,20 +508,34 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
+            [&](auto _) {
+              IValue value;
+              std::vector<std::reference_wrapper<const at::DataPtr>> dataPtrs;
+              std::tie(value, dataPtrs) = _(cb)(parentFut);
```
would I be correct that `_(cb)` converts the `cb` type to `IValueWithDataPtrs && (Future &)`?

lw replied:
I'm not sure about the details, but according to the docstring of `guts::if_constexpr` we need `_` in order to "trick" the compiler into delaying type-checking and thus avoid it complaining about the fact that `cb` has different types in the two branches. In practice, `_` is the identity function, so it should have no effect.

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]

lw committed:
Update on "Allow Future::then to return pre-extracted DataPtrs"


In CUDA mode, Future must inspect its value and extract DataPtrs. However some types are not supported, for example the C++/JIT custom classes, which include Message, which is widely used in RPC. Hence for these scenarios we allow the user to perform the custom DataPtr extraction on their own, and pass the pre-extracted DataPtrs.

Note that `markCompleted` already allowed users to pass in pre-extracted DataPtrs, hence this PR simply extends this possibility to the `then` method too.

Differential Revision: [D28474880](https://our.internmc.facebook.com/intern/diff/D28474880/)

[ghstack-poisoned]
